// @croc/webclient/src/lib/ui/menu/MenuPresenter.js
define(["require", "exports", "jquery", "core", "lib/ui/menu/MenuPresenterBase", "xcss!lib/ui/styles/menu.css"], function (require, exports, $, core, MenuPresenterBase) {
    "use strict";
    var MenuPresenter = (function (_super) {
        __extends(MenuPresenter, _super);
        /**
         * Menu presenter for action menus (based on buttons)
         * @constructs MenuPresenter
         * @extends MenuPresenterBase
         * @param options See MenuPresenterBase.prototype.defaultOptions
         * @param {Object} options.viewModel
         */
        function MenuPresenter(options) {
            var _this = this;
            options = MenuPresenterBase.mixOptions(options, MenuPresenter.defaultOptions);
            _this = _super.call(this, options) || this;
            return _this;
        }
        MenuPresenter.prototype._createRootEl = function (domElement) {
            return $("<div class='" + this.options.classes.root + "'>").appendTo(domElement);
        };
        MenuPresenter.prototype._createSubmenuContainer = function (item) {
            return $("<div class='" + this.options.classes.submenuContainer + "'></div>");
        };
        MenuPresenter.prototype._createItemEl = function (item) {
            var btn = $("<button type='button' class='" + this.options.classes.item + "' />");
            this._addItemCommonAttrs(item, btn);
            btn.button();
            return btn;
        };
        MenuPresenter.defaultOptions = {
            classes: {
                root: "x-menu",
                item: "x-menu-item",
                itemDefault: "x-menu-item-default",
                submenuContainer: "x-menu-item-dropdown"
            }
        };
        return MenuPresenter;
    }(MenuPresenterBase));
    core.ui.MenuPresenter = MenuPresenter;
    return MenuPresenter;
});
