// @croc/webclient/src/modules/transfer/ui/ExportPart.js
define(["require", "exports", "core", "lib/ui/handlebars/View", "xhtmpl!modules/transfer/ui/templates/ExportPart.hbs", "modules/transfer/Transfer", "i18n!modules/transfer/nls/resources"], function (require, exports, core, View, template, transfer, resourcesModule) {
    "use strict";
    /**
     * @enum
     */
    var ExportPartState;
    (function (ExportPartState) {
        ExportPartState[ExportPartState["connecting"] = 0] = "connecting";
        ExportPartState[ExportPartState["initial"] = 1] = "initial";
        ExportPartState[ExportPartState["uploading"] = 2] = "uploading";
        ExportPartState[ExportPartState["startingUp"] = 3] = "startingUp";
        ExportPartState[ExportPartState["importing"] = 4] = "importing";
        ExportPartState[ExportPartState["aborting"] = 5] = "aborting";
        ExportPartState[ExportPartState["aborted"] = 6] = "aborted";
        ExportPartState[ExportPartState["failed"] = 7] = "failed";
        ExportPartState[ExportPartState["suspended"] = 8] = "suspended";
        ExportPartState[ExportPartState["completed"] = 9] = "completed";
        ExportPartState[ExportPartState["disposed"] = 10] = "disposed";
    })(ExportPartState || (ExportPartState = {}));
    var ExportPart = (function (_super) {
        __extends(ExportPart, _super);
        function ExportPart(app, options) {
            var _this = this;
            options = ExportPart.mixOptions(options, ExportPart.defaultOptions);
            _this = _super.call(this, options) || this;
            _this.app = app;
            _this.client = _this.options.client || new transfer.ExportClient(app);
            return _this;
        }
        ExportPart.State = ExportPartState;
        ExportPart.defaultOptions = {
            template: template,
            scenario: undefined,
            client: undefined,
            commands: undefined,
            menu: { items: [
                    { name: "Abort", title: resourcesModule["transfer.cmd.abort"], order: 50 },
                    { name: "Close", title: resourcesModule["transfer.cmd.close"], order: 100 }
                ] }
        };
        return ExportPart;
    }(View));
    ExportPart.mixin(/** @lends ImportPart.prototype */ {
        defaultOptions: ExportPart.defaultOptions,
        /**
         * @observable-property {ImportPart#states}
         */
        state: core.lang.Observable.accessor("state"),
        /**
         * @observable-property {number}
         */
        progress: core.lang.Observable.accessor("progress")
    });
    return ExportPart;
});
