// @croc/webclient/src/lib/interop/.interop.types.ts
export const LoadRule = {
	remoteFirst: "remoteFirst" as "remoteFirst",
	localFirst: "localFirst" as "localFirst",
	localIfOffline: "localIfOffline" as "localIfOffline",
	cached: "cached" as "cached",
	remoteOnly: "remoteOnly" as "remoteOnly",
	localOnly: "localOnly" as "localOnly"
};
export type LoadRule = (typeof LoadRule)[keyof typeof LoadRule];

export const SaveTarget = {
	remoteFirst: "remoteFirst" as "remoteFirst",
	remoteOnly: "remoteOnly" as "remoteOnly",
	local: "local" as "local"
};
export type SaveTarget = (typeof SaveTarget)[keyof typeof SaveTarget];

export const SaveMode = {
	smart: "smart" as "smart",
	offline: "offline" as "offline",
	remoteOnly: "remoteOnly" as "remoteOnly"
};
export type SaveMode = (typeof SaveMode)[keyof typeof SaveMode];

export const AppCacheState = {
	uncached: "uncached" as "uncached",
	error: "error" as "error",
	cached: "cached" as "cached",
	noupdate: "noupdate" as "noupdate",
	obsolete: "obsolete" as "obsolete",
	updateready: "updateready" as "updateready"
};
export type AppCacheState = (typeof AppCacheState)[keyof typeof AppCacheState];
