// @croc/webclient/src/lib/ui/slick/SlickGridDataProvider.ts
import core = require("core");
import "lib/ui/slick/slick.bootstrap";

import lang = core.lang;
import ObservableExpression = lang.support.ObservableExpression;
import DependencyTracker = lang.support.DependencyTracker;
import { IList } from "ui/list/.list";
import { Violation } from "lib/validation";
import Options = SlickGridDataProvider.Options;
import RowsMap = SlickGridDataProvider.RowsMap;
import GroupingOptions = SlickGridDataProvider.GroupingOptions;
import NonDataRow = SlickGridDataProvider.NonDataRow;
import SlickGroup = SlickGridDataProvider.SlickGroup;
import SlickGroupTotals = SlickGridDataProvider.SlickGroupTotals;
import GridRow = SlickGridDataProvider.GridRow;
import GridRowMetadata = SlickGridDataProvider.GridRowMetadata;

interface GroupingInfo extends GroupingOptions {
	getterIsAFn: boolean;
	collapsed: any;
}

/**
 * ObservableExpression which also observes changing of id
 */
class ObservableExpressionTrackingId extends ObservableExpression {
	protected _track(source: any, args?: any[]|IArguments): DependencyTracker {
		let tracker: DependencyTracker = super._track(source, args);
		tracker.observe(source, "id"); //observe changing of id
		return tracker;
	}
}

class SlickGridDataItem {
	item: any;

	private _expr: ObservableExpression;
	private _loadingValue: any;
	private _vals: lang.Map<any>;

	/**
	 * @constructs SlickGridDataItem
	 * @param item
	 * @param getters
	 * @param exprOptions
	 * @param {Boolean} [trackId] Observe changing of id also
	 */
	constructor(item: any, getters: lang.Map<() => any>, exprOptions: ObservableExpression.Options, trackId?: boolean) {
		let that = this;

		that.item = item;
		if (lang.support.isNotLoaded(that.item)) {
			// NOTE: после загрузки может поменяться исходный объект
			that.item.load().done((loadedObj) => {
				that.item = loadedObj;
			});
		}

		// create an observable expression, which calculates values of all getters
		// NOTE: we're using class ObservableExpression directly to speed up
		let func = function SlickGridDataItemFunc(): lang.Map<any> {
			let caller = this,
				values: lang.Map<any> = {};
			lang.forEach(getters, function (getter: () => any, prop: string): void {
				try {
					values[prop] = getter.call(caller);
				} catch (ex) {
					values[prop] = exprOptions.errorValue;
				}
			});
			return values;
		};

		that._expr = !trackId ?
			new ObservableExpression(func, exprOptions) :
			new ObservableExpressionTrackingId(func, exprOptions);

		that._loadingValue = exprOptions.loadingValue;
	}

	dispose(): void {
		let that = this;
		that.invalidate();
		that._expr.dispose();
		that._expr = undefined;
	}

	invalidate(): void {
		this._vals = undefined;
	}

	get(prop: string): any {
		let that = this,
			values = that._vals;
		if (!values) {
			values = that._vals = that._expr.evaluate(that.item);
		}
		return values === that._loadingValue ? that._loadingValue : values[prop];
	}
}

class SlickGridDataProvider extends lang.Observable implements Slick.DataProvider<Slick.SlickData> {
	static defaultOptions: Options = {
		dataItemMetadataProvider: undefined,
		groupItemMetadataProvider: undefined,
		defaultGroupOptions: {} //see SlickObjectListDataPresenter.defaultGroupOptions
	};

	viewModel: IList;
	options: Options;

	private _items: SlickGridDataItem[];
	private _getters: lang.Map<() => any>;
	private _groupingInfos: GroupingInfo[];
	private _groups: SlickGroup[];
	private _toggledGroupsByLevel: lang.Map<number>[]; // lang.Map<0|1>[] - values are 0 or 1
	private _groupingDelimiter: string;
	private _rowsById: lang.Map<number>;
	private _groupRowsById: lang.Map<number>;
	private _groupTotalRowsById: lang.Map<number>;
	private _rows: GridRow[];
	/**
	 * Maps the index of the row (the same as in `_rows`) to the index as if grouping is off.
	 */
	private _ungroupedRows: number[];

	/**
	 * @constructs SlickGridDataProvider
	 * @extends Observable
	 * @param viewModel
	 * @param options
	 */
	constructor(viewModel: IList, options?: Options) {
		super();

		let that = this;
		that.viewModel = viewModel;			// viewModel with domain objects
		that._items = [];					// SlickGridDataItems by viewModel indexes
		that._getters = {};

		that.options = lang.append(options || {}, SlickGridDataProvider.defaultOptions);
		that._groupingInfos = [];			// group metadata
		that._groups = [];					// group objects
		that._toggledGroupsByLevel = [];	// group collapsed flag. usage: _toggledGroupsByLevel[level][groupingKey]
		that._groupingDelimiter = ":|:";	// grouping key delimiter.
		that._rowsById = null;				// data row number by id cache
		that._groupRowsById = null;			// group title row number by grouping key cache
		that._groupTotalRowsById = null;	// group totals row number by grouping key cache
		that._rows = null;					// current rows (Contains mix of SlickGridDataItems, group titles rows and group totals).

		// observe metadata changes
		that._getters["__itemMetadata"] = function (): GridRowMetadata {
			return that.options.dataItemMetadataProvider.getItemMetadata(this);
		};

		for (let col of that.viewModel.columns) {
			if (col.name) {
				that._getters[col.name] = function (): string {
					// NOTE: 'this' is a item of the list here, don't' use lambda
					return that.viewModel.getCellHtml(this, col);
				};
			}
		}

		that.viewModel.items.bind("change", that.onItemsChange, that);
		that.viewModel.selection.bind("change", that.onSelectionChange, that);
		that.viewModel.violations.bind("change", that.onViolationsChange, that);
	}

	/**
	 * Special value that represent loading cell
	 */
	loadingValue = lang.support.loadingValue;

	/**
	 * Special value that represent an invalid cell
	 */
	errorValue = {
		toString: function () { return "Error!"; }
	};

	dispose(): void {
		let that = this;
		that.invalidate();
		that.viewModel.items.unbind("change", null, that);
		that.viewModel.selection.unbind("change", null, that);
		that.viewModel.violations.unbind("change", null, that);
	}

	invalidate(): void {
		let that = this;

		that._items.forEach((item: SlickGridDataItem) => {
			if (item) {
				item.dispose();
			}
		});

		that._items.length = 0;
		that.clearRows();
	}

	/**
	 * Invalidate an item with the specified index
	 * @param viewModelIndex
	 * @returns {RowsMap} Rows to refresh
	 */
	invalidateItem(viewModelIndex: number): RowsMap {
		let that = this,
			item = that._items[viewModelIndex],
			rows: RowsMap;

		if (item) {
			item.invalidate();
		}

		if (that.isGroupingEnabled()) {
			// groups are possibly changed, clear rows cache
			that.clearRows();
			rows = { all: true };
		} else {
			rows = {};
			// no grouping => index in rows is equal to index in items
			rows[viewModelIndex] = true;
		}
		return rows;
	}

	getLength(): number {
		let that = this;
		//if grouping disabled, just proxy call to viewModel
		if (!that.isGroupingEnabled()){
			return that.getViewModelLength();
		}

		//else recalc rows and groups if necessary
		that.ensureRows();
		return that._rows.length;
	}

	getItem(rowIndex: number): GridRow;
	getItem<T extends SlickGridDataItem>(rowIndex: number): T;
	getItem<T extends SlickGroup>(rowIndex: number): T;
	getItem<T extends SlickGroupTotals>(rowIndex: number): T;
	getItem(rowIndex: number): GridRow {
		let that = this;
		//if grouping disabled, just proxy call to viewModel
		if (!that.isGroupingEnabled()) {
			return that.getItemFromViewModel(rowIndex);
		}

		//else recalc rows and groups if necessary and return item
		that.ensureRows();
		return that._rows[rowIndex];
	}

	getItemMetadata(rowIndex: number): GridRowMetadata {
		let that = this,
			item: GridRow = that.getItem(rowIndex);

		if (!item) { return {}; }

		// overrides for grouping rows
		if ((<SlickGroup>item).__group) {
			let getGroupRowMetadata = this._groupingInfos[(<SlickGroup>item).level].getGroupRowMetadata || that.options.groupItemMetadataProvider.getGroupRowMetadata;
			return getGroupRowMetadata.call(that.options.groupItemMetadataProvider, <SlickGroup>item);
			//return that.options.groupItemMetadataProvider.getGroupRowMetadata(<SlickGroup>item);
		}

		// overrides for totals rows
		if ((<SlickGroupTotals>item).__groupTotals) {
			let getTotalsRowMetadata = this._groupingInfos[(<SlickGroupTotals>item).group.level].getTotalsRowMetadata || that.options.groupItemMetadataProvider.getTotalsRowMetadata;
			return getTotalsRowMetadata.call(that.options.groupItemMetadataProvider, <SlickGroupTotals>item);
			//return that.options.groupItemMetadataProvider.getTotalsRowMetadata(<SlickGroupTotals>item);
		}

		// NOTE: we're calling "__itemMetadata" getter, which is from that.options.dataItemMetadataProvider.getItemMetadata
		let itemMeta = (<SlickGridDataItem>item).get("__itemMetadata");
		return itemMeta === that.loadingValue ? {} : itemMeta;
	}

	getItemIndex(item: GridRow): number {
		let that = this;
		if (!that.isGroupingEnabled()) {
			return that.viewModel.items.indexOf((<SlickGridDataItem>item).item);
		}

		that.ensureRows();
		if ((<SlickGroupTotals>item).__groupTotals){
			return that._groupTotalRowsById[(<SlickGroupTotals>item).group.groupingKey];
		}
		if ((<SlickGroup>item).__group){
			return that._groupRowsById[(<SlickGroup>item).groupingKey];
		}
		return that._rowsById[(<SlickGridDataItem>item).item.id];
	}

	getModelItemIndex(viewModelItem: any): number {
		let that = this;
		if (!that.isGroupingEnabled()) {
			return that.viewModel.items.indexOf(viewModelItem);
		}

		that.ensureRows();
		return that._rowsById[viewModelItem.id];
	}

	isItemSelectable(rowIndex: number): boolean {
		let row = this.getItem(rowIndex),
			item = row && (row as SlickGridDataItem).item;
		// NOTE: group/totals rows have no item
		return !!(item && this.viewModel.isItemSelectable(item));
	}

	/**
	 * Returns the index of the row as if grouping is off.
	 * @param rowIndex
	 * @returns {number}
	 */
	getUngroupedIndex(rowIndex: number): number {
		let that = this;

		if (!that.isGroupingEnabled()) {
			return rowIndex;
		}

		that.ensureRows();
		return that._ungroupedRows[rowIndex];

		// NOTE: Groups are sorted themselves, therefore the order of items in viewModel can be different
		// from the order of rows. So the commented code is incorrect.
		//let row = that._rows[rowIndex];
		//return that.viewModel.items.indexOf((<SlickGridDataItem>row).item);
	}

	getGrouping(): GroupingInfo[] {
		return this._groupingInfos;
	}

	setGrouping(groupingInfo: GroupingOptions | GroupingOptions[]): void {
		let that = this;

		that._groups = [];
		that._toggledGroupsByLevel = [];
		that.clearRows();

		groupingInfo = groupingInfo || [];
		that._groupingInfos = lang.isArray(groupingInfo) ? <GroupingInfo[]>groupingInfo : [<GroupingInfo>groupingInfo];

		for (let i = 0; i < that._groupingInfos.length; i++) {
			let gi = that._groupingInfos[i] = lang.extend({}, that.options.defaultGroupOptions, that._groupingInfos[i]);
			gi.getterIsAFn = typeof gi.getter === "function";

			that._toggledGroupsByLevel[i] = {};
		}

		that.invalidate();
	}

	/**
	 * @param args Either a Slick.Group's "groupingKey" property, or a
	 * variable argument list of grouping values denoting a unique path to the row.  For
	 * example, calling collapseGroup('high', '10%') will collapse the '10%' subgroup of
	 * the 'high' setGrouping.
	 */
	collapseGroup(...args: string[]): void {
		let that = this,
			arg0 = args[0];
		if (args.length === 1 && arg0.indexOf(that._groupingDelimiter) !== -1) {
			that._toggleGroup(arg0.split(that._groupingDelimiter).length - 1, arg0, true);
		} else {
			that._toggleGroup(args.length - 1, args.join(that._groupingDelimiter), true);
		}
	}

	/**
	 * @param args Either a Slick.Group's "groupingKey" property, or a
	 * variable argument list of grouping values denoting a unique path to the row.  For
	 * example, calling expandGroup('high', '10%') will expand the '10%' subgroup of
	 * the 'high' setGrouping.
	 */
	expandGroup(...args: string[]): void {
		let that = this,
			arg0 = args[0];
		if (args.length === 1 && arg0.indexOf(that._groupingDelimiter) !== -1) {
			that._toggleGroup(arg0.split(that._groupingDelimiter).length - 1, arg0, false);
		} else {
			that._toggleGroup(args.length - 1, args.join(that._groupingDelimiter), false);
		}
	}

	protected _extractGroups(rows: SlickGridDataItem[], parentGroup?: SlickGroup): SlickGroup[] {
		let that = this,
			groups: SlickGroup[] = [],
			groupsByVal: lang.Map<SlickGroup> = {},
			level = parentGroup ? parentGroup.level + 1 : 0,
			gi: GroupingInfo = that._groupingInfos[level],
			comparer = <(x: SlickGroup, y: SlickGroup) => number> gi.comparer;

		for (let val of gi.predefinedValues) {
			let key = that._getGroupingKey(val),
				group = groupsByVal[key];
			if (!group) {
				group = new Slick.Group<SlickGridDataItem>();
				group.value = val;
				group.level = level;
				group.groupingKey = (parentGroup ? parentGroup.groupingKey + that._groupingDelimiter : "") + key;
				groups[groups.length] = group;
				groupsByVal[key] = group;
			}
		}

		for (let r of rows) {
			let val = that._getGroupingValue(r.item, gi),
				key = that._getGroupingKey(val),
				group = groupsByVal[key];
			if (!group) {
				group = new Slick.Group<SlickGridDataItem>();
				group.value = val;
				group.level = level;
				group.groupingKey = (parentGroup ? parentGroup.groupingKey + that._groupingDelimiter : "") + key;
				groups[groups.length] = group;
				groupsByVal[key] = group;
			}
			group.rows[group.count++] = r;
		}

		if (level < that._groupingInfos.length - 1) {
			for (let group of groups) {
				group.groups = that._extractGroups(group.rows, group);
			}
		}

		if (comparer) {
			groups = lang.sort(groups, comparer);
		}

		return groups;
	}

	protected _getGroupingValue(item: any, gi: GroupingInfo): any {
		return gi.getterIsAFn ?
			(gi.getter as (i: any) => any)(item) :
			lang.get(item, gi.getter as string);
	}

	protected _getGroupingKey(val: any): string {
		let key = (val && val.hasOwnProperty("id")) ? val.id : val; //if value - domain object, use id
		return "" + key;
	}

	protected _calculateGroupTotals(group: SlickGroup): void {
		let that = this,
			gi = that._groupingInfos[group.level],
			isLeafLevel = (group.level === that._groupingInfos.length),
			totals = new Slick.GroupTotals<SlickGridDataItem>(),
			agg,
			idx = gi.aggregators.length,
			rowItemGetter = function (r) { return r.item; };
		while (idx--) {
			agg = gi.aggregators[idx];
			agg.init();
			agg.accumulate((!isLeafLevel && gi.aggregateChildGroups) ? group.groups : group.rows.map(rowItemGetter));
			agg.storeResult(totals);
		}
		totals.group = group;
		group.totals = totals;
	}

	protected _calculateTotals(groups: SlickGroup[], level: number = 0): void {
		let that = this,
			gi = that._groupingInfos[level],
			idx = groups.length,
			g: SlickGroup;
		while (idx--) {
			g = groups[idx];

			if (g.collapsed && !gi.aggregateCollapsed) {
				continue;
			}

			// Do a depth-first aggregation so that parent setGrouping aggregators can access subgroup totals.
			if (g.groups) {
				that._calculateTotals(g.groups, level + 1);
			}

			if (gi.aggregators.length && (
				gi.aggregateEmpty || g.rows.length || (g.groups && g.groups.length))) {
				that._calculateGroupTotals(g);
			}
		}
	}

	protected _finalizeGroups(groups: SlickGroup[], level: number = 0): void {
		let that = this,
			gi = that._groupingInfos[level],
			groupCollapsed = gi.collapsed,
			toggledGroups = that._toggledGroupsByLevel[level],
			idx = groups.length,
			g;

		while (idx--) {
			g = groups[idx];
			g.collapsed = groupCollapsed ^ toggledGroups[g.groupingKey];
			g.title = gi.formatter ? gi.formatter(g) : g.value;

			if (g.groups) {
				that._finalizeGroups(g.groups, level + 1);
				// Let the non-leaf setGrouping rows get garbage-collected.
				// They may have been used by aggregates that go over all of the descendants,
				// but at this point they are no longer needed.
				g.rows = [];
			}
		}
	}

	protected _flattenGroupedRows(groups: SlickGroup[], level: number = 0): GridRow[] {
		let that = this,
			gi = that._groupingInfos[level],
			groupedRows: GridRow[] = [],
			gl = 0;
		for (let g of groups) {
			groupedRows[gl++] = g;

			if (!g.collapsed) {
				let rows = g.groups ? that._flattenGroupedRows(g.groups, level + 1) : g.rows;
				for (let r of rows) {
					groupedRows[gl++] = r;
				}
			}

			if (g.totals && gi.displayTotalsRow && (!g.collapsed || gi.aggregateCollapsed)) {
				groupedRows[gl++] = g.totals;
			}
		}
		if (groups.length === 1 && groups[0].value === undefined) {
			// there is the only one unnamed group - hide it
			groupedRows = groupedRows.slice(1);
		}
		return groupedRows;
	}

	protected getItemFromViewModel(viewModelIndex: number): SlickGridDataItem {
		let that = this,
			viewModelItem,
			item = that._items[viewModelIndex];
		if (!item) {
			viewModelItem = that.viewModel.items.get(viewModelIndex);
			item = new SlickGridDataItem(
				viewModelItem,
				that._getters,
				{
					onchange() { that.onItemChange(viewModelIndex); },
					loadingValue: that.loadingValue,
					errorValue: that.errorValue
				},
				// NOTE: If grouping is enable, id of the item is used in _rowsById dictionary.
				// When id changes, onItemChange callback should be called to reset _rowsById.
				// So we must observe changing of id also - specify `trackId` flag
				that.isGroupingEnabled()
			);
			that._items[viewModelIndex] = item;
		}
		return item;
	}

	protected getViewModelLength(): number {
		return this.viewModel.items.count();
	}

	protected _toggleGroup(level: number, groupingKey: string, collapse: boolean): void {
		let that = this,
			row = that._groupRowsById[groupingKey];

		that._toggledGroupsByLevel[level][groupingKey] = that._groupingInfos[level].collapsed ^ <any>collapse;

		that.trigger("group.collapsing", that, {
			row: row,
			group: that._rows[row] as SlickGroup,
			collapse: collapse
		});

		that.invalidate();
		that.notifyRowsChanged();
	}

	protected isGroupingEnabled(): boolean {
		return !!this._groupingInfos.length;
	}

	protected onItemChange(viewModelIndex: number): void {
		let rows = this.invalidateItem(viewModelIndex);
		this.notifyRowsChanged(rows);
	}

	protected onItemsChange(): void {
		this.invalidate();
		this.notifyRowsChanged();
	}

	protected onSelectionChange(sender, args: lang.ObservableCollectionChangeArgs<any>): void {
		let that = this;
		if (!that.isGroupingEnabled()) { return; }

		if (args && args.added) {
			that.ensureRows();

			// expand groups to ensure selected items are visible
			let expanded: boolean;
			for (let item of args.added) {
				if (that._rowsById[item.id] >= 0) { continue; } // row is already expanded

				let level = 0,
					groupKey = "";
				for (let gi of that._groupingInfos) {
					let val = that._getGroupingValue(item, gi),
						key = that._getGroupingKey(val);

					if (groupKey) {
						groupKey += that._groupingDelimiter;
					}
					groupKey += key;

					that._toggledGroupsByLevel[level][groupKey] = that._groupingInfos[level].collapsed ^ 0;
					level++;
				}

				expanded = true;
			}

			// invalide rows (only if at least one group was expanded)
			if (expanded) {
				that.invalidate();
				that.notifyRowsChanged();
			}
		}
	}

	protected onViolationsChange(sender, args: lang.ObservableCollectionChangeArgs<Violation>): void {
		let violations: Violation[] = [];

		if (args) {
			if (args.added) {
				violations = violations.concat(args.added);
			}
			if (args.removed) {
				violations = violations.concat(args.removed);
			}
		}

		// invalidate rows to update metadata
		let rows: RowsMap = {};
		violations.forEach((violation: Violation) => {
			if (!violation.object) { return; }

			let row = this.getModelItemIndex(violation.object);
			if (row >= 0 && !rows[row]) {
				rows[row] = true;
				let dataItem = this.getItem(row) as SlickGridDataItem;
				if (lang.isFunction(dataItem.invalidate)) {
					dataItem.invalidate();
				}
			}
		});

		if (!lang.isEmpty(rows)) {
			this.notifyRowsChanged(rows);
		}
	}

	protected clearRows(): void {
		let that = this;

		//clear indexes
		that._ungroupedRows = null;
		that._rowsById = null;
		that._groupRowsById = null;
		that._groupTotalRowsById = null;

		//and rows
		that._rows = null;
	}

	protected ensureRows(): void {
		let that = this,
			dataRows: SlickGridDataItem[];

		if (that._rows) { return; }

		//getting items
		dataRows = [];
		for (let i = 0, l = that.getViewModelLength(); i < l; i++) {
			dataRows.push(that.getItemFromViewModel(i));
		}

		// clear caches and indexes
		that.clearRows();

		// clear and recreate groups
		that._groups = [];
		if (that._groupingInfos.length) {
			that._groups = that._extractGroups(dataRows);
			if (that._groups.length) {
				that._calculateTotals(that._groups);
				that._finalizeGroups(that._groups);
				that._rows = that._flattenGroupedRows(that._groups);
			}
		}

		that._rows = that._rows || dataRows;

		// build cache
		that._rowsById = {};
		that._groupRowsById = {};
		that._groupTotalRowsById = {};
		that._ungroupedRows = [];
		for (let i = 0, l = that._rows.length, n = 0; i < l; i++) {
			let row: GridRow = that._rows[i];
			if ((<SlickGroupTotals>row).__groupTotals) {
				that._groupTotalRowsById[(<SlickGroupTotals>row).group.groupingKey] = i;
			} else if ((<SlickGroup>row).__group) {
				that._groupRowsById[(<SlickGroup>row).groupingKey] = i;
				if ((<SlickGroup>row).collapsed) { n += (<SlickGroup>row).count; }
			} else {
				that._rowsById[(<SlickGridDataItem>row).item.id] = i;
				that._ungroupedRows[i] = n++;
			}
		}
	}

	protected notifyRowsChanged(rows?: RowsMap): void {
		let that = this;

		// NOTE: if grouping is enabled all rows should be updated
		if (!rows || that.isGroupingEnabled()) {
			rows = { all: true };
		}
		that.trigger("rows.change", that, { rows: rows });
	}
}

namespace SlickGridDataProvider {
	export interface Options {
		dataItemMetadataProvider: {
			getItemMetadata(item: any): GridRowMetadata;
		};
		groupItemMetadataProvider: Slick.Data.GroupItemMetadataProvider<Slick.SlickData>;
		defaultGroupOptions: Slick.Data.GroupingOptions<Slick.SlickData>;
	}

	export interface RowsMap {
		all?: boolean;
		[row: number]: boolean;
	}

	export interface GroupingOptions extends Slick.Data.GroupingOptions<Slick.SlickData> {
		getGroupRowMetadata?: (item: any) => {};
		getTotalsRowMetadata?: (item: any) => {};
	}

	export interface DataRow extends SlickGridDataItem {
	}

	export interface NonDataRow extends Slick.NonDataRow {
		__nonDataRow?: boolean;
	}

	export interface SlickGroup extends NonDataRow, Slick.Group<SlickGridDataItem> {
		__group?: boolean;
	}

	export interface SlickGroupTotals extends NonDataRow, Slick.GroupTotals<SlickGridDataItem> {
		__groupTotals?: boolean;
	}

	export type GridRow = SlickGridDataItem | SlickGroup | SlickGroupTotals;

	export type GridRowMetadata = Slick.RowMetadata<GridRow>;

	export interface GroupCollapsingEventArgs {
		row: number;
		group: SlickGroup,
		collapse: boolean;
	}
}

SlickGridDataProvider.mixin({ defaultOptions: SlickGridDataProvider.defaultOptions });

export = SlickGridDataProvider;
