// @croc/webclient/src/modules/reporting/module-reporting.ts
import core = require("core");
import ReportPartBase = require("./ui/ReportPartBase");
import ReportPagePart = require("./ui/ReportPagePart");
import ReportPagePartPresenter = require("./ui/ReportPagePartPresenter");
import ReportPart = require("./ui/ReportPart");
import ReportPartPresenter = require("./ui/ReportPartPresenter");
import ReportPresenterBase = require("./ui/ReportPresenterBase");
import ObjectList = require("./ui/ObjectListExportable");
import resources = require("i18n!lib/nls/resources");
import moduleResources = require("i18n!modules/reporting/nls/resources");

// extend common resources
core.lang.forEach(moduleResources, function (value, key) {
	resources[key] = value;
});

core.createModule("reporting", function (app: core.Application, options) {
	options =  options || {};
	let formatsConfig = options.formats,
		menuItem,
		exportMenuItems,
		exportMenuItemsWOHtml;

	if (formatsConfig && formatsConfig.length > 0) {
		exportMenuItems = formatsConfig.map((format) => {
			return {
				name: "Export-" + format.name,
				title: format.name,
				commandName: "Export",
				params: { format: format.name, mime: format.mime }
			};
		});

		exportMenuItemsWOHtml = exportMenuItems
			.filter((item) => { return item.params.format !== "HTML"; } );

		// ObjectList menu:
		menuItem = {
			name: "Export",
			title: resources["objectList.exportTo"],
			icon: "export",
			order: 40,
			items: exportMenuItemsWOHtml,
			command: null
		};
		ObjectList.defaultMenus.List.items.push(menuItem);
		ObjectList.defaultMenus.EditableList.items.push(menuItem);

		// ReportPart menu:
		menuItem = {
			name: "Export",
			title: resources["reportPart.openWith"],
			icon: "export",
			order: 30,
			items: exportMenuItems,
			command: null
		};
		ReportPart.defaultMenu.items.push(menuItem);

		// ReportPagePart menu:
		menuItem = core.lang.append({
			items: exportMenuItemsWOHtml
		}, menuItem);
		ReportPagePart.defaultMenu.items.push(menuItem);

		core.registerPart("ReportPart", function (options) {
			return new ReportPart(app, options);
		});
	}
});

let reporting = {
	Application: null,
	ReportPartBase: ReportPartBase,
	ReportPart: ReportPart,
	ReportPagePart: ReportPagePart,
	ReportPresenterBase: ReportPresenterBase,
	ReportPagePartPresenter: ReportPagePartPresenter,
	ReportPartPresenter: ReportPartPresenter
};
core.reporting = reporting;

export = reporting;
