// @croc/webclient/src/modules/reporting/module-reporting.js
define(["require", "exports", "core", "./ui/ReportPartBase", "./ui/ReportPagePart", "./ui/ReportPagePartPresenter", "./ui/ReportPart", "./ui/ReportPartPresenter", "./ui/ReportPresenterBase", "./ui/ObjectListExportable", "i18n!lib/nls/resources", "i18n!modules/reporting/nls/resources"], function (require, exports, core, ReportPartBase, ReportPagePart, ReportPagePartPresenter, ReportPart, ReportPartPresenter, ReportPresenterBase, ObjectList, resources, moduleResources) {
    "use strict";
    // extend common resources
    core.lang.forEach(moduleResources, function (value, key) {
        resources[key] = value;
    });
    core.createModule("reporting", function (app, options) {
        options = options || {};
        var formatsConfig = options.formats, menuItem, exportMenuItems, exportMenuItemsWOHtml;
        if (formatsConfig && formatsConfig.length > 0) {
            exportMenuItems = formatsConfig.map(function (format) {
                return {
                    name: "Export-" + format.name,
                    title: format.name,
                    commandName: "Export",
                    params: { format: format.name, mime: format.mime }
                };
            });
            exportMenuItemsWOHtml = exportMenuItems
                .filter(function (item) { return item.params.format !== "HTML"; });
            // ObjectList menu:
            menuItem = {
                name: "Export",
                title: resources["objectList.exportTo"],
                icon: "export",
                order: 40,
                items: exportMenuItemsWOHtml,
                command: null
            };
            ObjectList.defaultMenus.List.items.push(menuItem);
            ObjectList.defaultMenus.EditableList.items.push(menuItem);
            // ReportPart menu:
            menuItem = {
                name: "Export",
                title: resources["reportPart.openWith"],
                icon: "export",
                order: 30,
                items: exportMenuItems,
                command: null
            };
            ReportPart.defaultMenu.items.push(menuItem);
            // ReportPagePart menu:
            menuItem = core.lang.append({
                items: exportMenuItemsWOHtml
            }, menuItem);
            ReportPagePart.defaultMenu.items.push(menuItem);
            core.registerPart("ReportPart", function (options) {
                return new ReportPart(app, options);
            });
        }
    });
    var reporting = {
        Application: null,
        ReportPartBase: ReportPartBase,
        ReportPart: ReportPart,
        ReportPagePart: ReportPagePart,
        ReportPresenterBase: ReportPresenterBase,
        ReportPagePartPresenter: ReportPagePartPresenter,
        ReportPartPresenter: ReportPartPresenter
    };
    core.reporting = reporting;
    return reporting;
});
