// @croc/webclient/src/lib/ui/editor/ObjectWizardPresenter.js
define(["require", "exports", "core", "lib/ui/editor/ObjectWizardPresenterBase", "xhtmpl!lib/ui/templates/Wizard.hbs", "xcss!lib/ui/styles/objectWizard.css"], function (require, exports, core, ObjectWizardPresenterBase, wizardTemplate) {
    "use strict";
    var ObjectWizardPresenter = (function (_super) {
        __extends(ObjectWizardPresenter, _super);
        /**
         * @constructs ObjectWizardPresenter
         * @extends ObjectWizardPresenterBase
         * @param {Object} [options]
         */
        function ObjectWizardPresenter(options) {
            var _this = this;
            options = ObjectWizardPresenter.mixOptions(options, ObjectWizardPresenter.defaultOptions);
            _this = _super.call(this, options) || this;
            return _this;
        }
        ObjectWizardPresenter.prototype.doRender = function (domElement) {
            var that = this;
            _super.prototype.doRender.call(this, domElement);
            that.breadcrumbs = that.container.find(".x-wizard-top-nav > .breadcrumb");
        };
        ObjectWizardPresenter.prototype.onReady = function () {
            var that = this, eventPublisher = that.viewModel && that.viewModel.app && that.viewModel.app.eventPublisher;
            if (eventPublisher && that.options.affixNavigation) {
                if (that.$topnav) {
                    eventPublisher.publish("ui.affix.remove_element", {
                        element: that.$topnav
                    });
                }
                that.$topnav = that.container.find(".x-wizard-top-nav");
                eventPublisher.publish("ui.affix.add_element", {
                    element: that.container.find(".x-wizard-top-nav"),
                    controlledBy: that.container.find(".x-editor-pages"),
                    affixTo: "top"
                });
            }
            _super.prototype.onReady.call(this);
        };
        ObjectWizardPresenter.prototype.unload = function () {
            var that = this, vm = that.viewModel;
            if (that.$topnav) {
                vm.app.eventPublisher.publish("ui.affix.remove_element", {
                    element: that.$topnav
                });
            }
            _super.prototype.unload.apply(that, arguments);
        };
        ObjectWizardPresenter.defaultOptions = {
            template: wizardTemplate
        };
        return ObjectWizardPresenter;
    }(ObjectWizardPresenterBase));
    core.ui.ObjectWizardPresenter = ObjectWizardPresenter;
    return ObjectWizardPresenter;
});
