// @croc/webclient/src/lib/ui/pe/peLoadableMixin.js
define(["require", "exports", "lib/core.lang"], function (require, exports, lang) {
    "use strict";
    var State = {
        initial: "initial",
        loading: "loading",
        loaded: "loaded",
        failed: "failed",
        disposed: "disposed"
    };
    var peLoadableMixin = (function () {
        function peLoadableMixin() {
        }
        peLoadableMixin.prototype._onDataLoading = function (args) {
            var that = this;
            // block input in lookup field (if exist)
            that.state(State.loading);
            that.onDataLoading(args);
            that._renderBeginLoading();
        };
        peLoadableMixin.prototype._renderBeginLoading = function () { };
        peLoadableMixin.prototype._onDataLoaded = function (args) {
            var that = this;
            that.onDataLoaded(args);
            that.lastError = null;
            that.isDataLoaded = true;
            that.state(State.loaded);
            that._renderEndLoading();
            that._setItems(args.items);
        };
        peLoadableMixin.prototype._renderEndLoading = function () { };
        peLoadableMixin.prototype._onFailed = function (error) {
            var that = this;
            that.lastError = error;
            that.isDataLoaded = false;
            that.state(State.failed);
            that._renderEndLoading();
            that.renderError(error);
        };
        peLoadableMixin.State = State;
        __decorate([
            lang.decorators.observableAccessor()
        ], peLoadableMixin.prototype, "state");
        return peLoadableMixin;
    }());
    return peLoadableMixin;
});
